Second Design Review for new DFW API

madhephaestus opened the pull request:
I have made the changes suggested by @AustinShalit and made a new PR for discussion.

madhephaestus committed:
adding Kevin Harrington to the authors page

madhephaestus committed:
API for constructor changed

The Constructor takes the serial port number, but in the comments it
notes that Serial1 is the only valid one. THis has been changed to only
access Serial1 and removed the parameter.

madhephaestus committed:
begin refactoring

The competition code is being switched over to a function pointer model.

madhephaestus committed:
Partial new packet parser working

madhephaestus committed:
Switching back to original parser

madhephaestus committed:
fixing the start button detect

madhephaestus committed:
BAsic competition state machine working

madhephaestus committed:
DWF v2 is ready for design review

madhephaestus committed:
Adding DFW pointer to lambdas

The functions that the user defines, now include the passing of the DFW
object from the fundtion call to the user function. THe lets the user
functions to be in isolated files and avoids code clutter. It also
provides an example to the users of how to pass objects around between
sections of code.

madhephaestus committed:
Updating API and packet parsing

The packet parsing is done in line rather than depending on the find
functions.

madhephaestus committed:
Move the update function to a private method

madhephaestus committed:
fix the polarity of button presses

madhephaestus committed:
Add an additional state for waiting for teleop to start

The start button needs to be hit a second time to kick the framework
into teleop mode.

madhephaestus committed:
Adding a status flash

The debug led will now flash when communication is working, and stay
solid on if it fails.

madhephaestus committed:
test code should check for powerup state

AustinShalit commented on DFW.cpp:
```diff
@@ -10,298 +10,325 @@
+		return false;
 	}
-	return (startB);
+	return true;
```
Can we return the comparison directly? i.e.: `return (byteBu[1] & 0b00000010) == 0`

AustinShalit commented on examples/DFWRBE1001Template/DFWRBE1001Template.ino:
```diff
@@ -1,90 +1,19 @@
+ time in seconds that each set of code should run. IE
+ autonomous(20); will run for 20 seconds after the transmitter is
+ turned on.The code will not start until the controller is turned on, connected, and start is pressed.
+ There are print statements commented out that were used to test */
```
This ¶ should probably be rewritten

AustinShalit commented on examples/DFWRBE1001Template/DFWRBE1001Template.ino:
```diff
@@ -1,90 +1,19 @@
+DFW dfw(ledpindebug, autonomous, teleop); // Instantiates the DFW object and setting the debug pin. The debug pin will be set high if no communication is seen after 2 seconds
+void setup() {
+	Serial.begin(9600); // Serial output begin. Only needed for debug
+	dfw.begin(9600); // Serial1 output begin for DFW library. Buad and port #."Serial1 only"
```
Could we provide a function that defaults to 9600? Then the code would be `dfw.begin()` and thus hide the baud from the user.

AustinShalit commented on examples/DFWRBE1001Template/teleop.cpp:
```diff
@@ -0,0 +1,10 @@
+	Serial.print(time);
+	Serial.print(" right joystick: ");
+	Serial.print(dfw.joystickrv());
+	Serial.print(" left joystick: ");
```
Instead of prepending the string with a space, do we want to use `\t`?  That way the spacing will be more consistent with changing numbers.

AustinShalit commented on examples/DFWRBE1001Template/teleop.cpp:
```diff
@@ -0,0 +1,10 @@
+	Serial.print(" right joystick: ");
+	Serial.print(dfw.joystickrv());
+	Serial.print(" left joystick: ");
+	Serial.print(dfw.joysticklv());
```
Should be `println`?

madhephaestus committed:
Communications tied to 9600 and single line returns for buttons

The buttons now return the result of the mask check directly. In
addition all of the initialization code loads using 9600 only.

madhephaestus committed:
tab separated prints

madhephaestus committed:
updated comments

madhephaestus committed:
spelling
